keps/sig-apps/4443-configurable-pod-failure-policy-reasons/README.md

```diff
@@ -0,0 +1,955 @@
+
+<!--
+Focus not just on happy cases, but primarily on more pathological cases
+(e.g. probes taking a minute instead of milliseconds, failed pods consuming resources, etc.).
+If any of the resources can be exhausted, how this is mitigated with the existing limits
+(e.g. pods per node) or new limits added by this KEP?
+
+Are there any tests that were run/should be run to understand performance characteristics better
+and validate the declared limits?
+-->
+No
+
+### Troubleshooting
+
+<!--
+This section must be completed when targeting beta to a release.
+
+For GA, this section is required: approvers should be able to confirm the
+previous answers based on experience in the field.
+
+The Troubleshooting section currently serves the `Playbook` role. We may consider
+splitting it into a dedicated `Playbook` document (potentially with some monitoring
+details). For now, we leave it here.
+-->
+
+###### How does this feature react if the API server and/or etcd is unavailable?
+
+###### What are other known failure modes?
+
+<!--
+For each of them, fill in the following information by copying the below template:
+  - [Failure mode brief description]
+    - Detection: How can it be detected via metrics? Stated another way:
+      how can an operator troubleshoot without logging into a master or worker node?
+    - Mitigations: What can be done to stop the bleeding, especially for already
+      running user workloads?
+    - Diagnostics: What are the useful log messages and their required logging
+      levels that could help debug the issue?
+      Not required until feature graduated to beta.
+    - Testing: Are there any tests for failure mode? If not, describe why.
+-->
+
+###### What steps should be taken if SLOs are not being met to determine the problem?
+
+## Implementation History
+
+<!--
+Major milestones in the lifecycle of a KEP should be tracked in this section.
+Major milestones might include:
+- the `Summary` and `Motivation` sections being merged, signaling SIG acceptance
+- the `Proposal` section being merged, signaling agreement on a proposed design
+- the date implementation started
+- the first Kubernetes release where an initial version of the KEP was available
+- the version of Kubernetes where the KEP graduated to general availability
+- when the KEP was retired or superseded
+-->
+- KEP Published: 02/05/2024
+
+## Drawbacks
+
+None
+
+## Alternatives
+
+Rather than having the user specify a custom reason via the proposed `SetConditionReason` field, which will require validation and an additional step
+in the user-experience, instead the Job controller could automatically generate reasons, using a determnistic
+format which depends on if the podFailurePolicy was triggered by the `onPodConditions` or `onContainerExitCodes`.
+
+For example, in the case of container exit codes, we could append the container exit code to
+the default reason (e.g., `PodFailurePolicy-ExitCode3`). For `onPodConditions`, it could be the pod condition
```

Comment:
@soltysh are you supportive of this approach? I am ok with this too. What do you think @alculquicondor?

Reply:
I'm not quite sure... Users need to go into the documentation to understand how to write it in a JobSet spec, and it could be error prone.

Also, what to do if the failure policy matches a set or range of exit codes?

The proposed approach seems easier to use IMO and it kind of matches the pattern of writing label selectors in a Deployment, matching the pod labels.

Reply:
> I'm not quite sure... Users need to go into the documentation to understand how to write it in a JobSet spec, and it could be error prone.
> 
> Also, what to do if the failure policy matches a set or range of exit codes?
> 
> The proposed approach seems easier to use IMO and it kind of matches the pattern of writing label selectors in a Deployment, matching the pod labels.

I agree. However, I am not strongly opposed to the alternative highlighted here, I just think the user experience for configuring a JobSet failure policy to utilize the condition reason added by a PodFailurePolicy would be more confusing and error prone, which is not ideal.